### ISSUE TITLE
[$250] [HOLD for payment 2024-04-05] Workspace - Not here page shows up when opening message in thread in deleted workspace

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#38183]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.51-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to workspace chat
3. Send a message
4. Delete the workspace
5. Return to the workspace chat
6. Hover over the message in Step 3 > Reply in thread

## Expected Result:
The message will be opened in archived thread

## Actual Result:
Not here page shows up when opening message in thread in deleted workspace

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/205edb3e-72a3-4719-9f32-4542c18d1501

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0129fa34eb79923fe5</li>
        <li>Upwork Job ID: 1777369739173396480</li>
        <li>Last Price Increase: 2024-04-08</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp
CC @quinthar

**Comment by lanitochka17** [comment]
@blimpich  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
Production: 


https://github.com/Expensify/App/assets/78819774/ec72f738-ce0f-405e-8b7f-f41eb8715a5b

**Comment by blimpich** [comment]
@GandalfGwaihir thank you for identifying the PR that caused this regression! However we generally regard deploy blockers to be the responsibility of who created them. So I would expect @VickyStash and @aimane-chnaif to handle this, as part of them completing https://github.com/Expensify/App/issues/25216.

Demoting this since its not a particularly bad bug, and is a more niche scenario.

**Comment by blimpich** [comment]
Hmmm, for some reason GitHub won't let me assign @VickyStash to this issue. Either way, @VickyStash @aimane-chnaif can you handle this issue as a follow up to the typescript conversion?

**Comment by VickyStash** [comment]
@blimpich Sure, on my way to prepare follow-up PR!

**Comment by VickyStash** [comment]
Btw, the suggested proposal didn't work for me, I still see the same message.
And for me, it looks like the problem is in the first part of the `shouldShowNotFoundPage` variable condition, cause that's what turning to `true` for me.
I'm continuing to investigate...

**Comment by nkdengineer** [comment]
## Proposal
 
### Please re-state the problem that we are trying to solve in this issue.
- WS switcher resets to Expensify when creating a room under the same WS
### What is the root cause of that problem?
- When replying in chat, we call ```navigateToAndOpenChildReport```
- In that function, there is logic that navigate user to thread:
https://github.com/Expensify/App/blob/91d7eb22b3cf0e690271a1313bf8016d03cb3ef7/src/libs/actions/Report.ts#L787-L788
- There is the case that the report data is not stored successfully before navigating to it, so that leads to the bug
### What changes do you think we should make in order to solve the problem?
- We need to make sure the ```report``` in [here](https://github.com/Expensify/App/blob/d482a25fc9e9f37990c27c1d5906227eb6204cf2/src/libs/Navigation/Navigation.ts#L60) is always fetched before calling ```originalDismissModalWithReport```.
- We can do it by creating a ```isReadyPromise``` in ReportUtils like the one below:
```
let waitingReportID: string = ''
let resolveIsReadyPromise: (args?: unknown[]) => void;
let isReadyPromise =(reportID: string)=> new Promise((resolve) => {
    waitingReportID = reportID
    resolveIsReadyPromise = resolve;
});

function checkRequiredData() {
    const targetReport = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${waitingReportID}`] ?? {};
    if(targetReport?.reportID){
        resolveIsReadyPromise();
    }
   
}
Onyx.connect({
    key: ONYXKEYS.COLLECTION.REPORT,
    waitForCollectionCallback: true,
    callback: (value) => {
        allReports = value
        checkRequiredData()
    },
});
```

- Then update:
https://github.com/Expensify/App/blob/d482a25fc9e9f37990c27c1d5906227eb6204cf2/src/libs/Navigation/Navigation.ts#L60-L61
to:
```
    const report = getReport(reportID);
   if(!isEmpty(report)){
        originalDismissModalWithReport({reportID, ...report}, ref);
        return
}
    ReportUtils.isReadyPromise(reportID).then(()=>{
        const report = getReport(reportID);
        originalDismissModalWithReport({reportID, ...report}, ref);
    })
```
### What alternative solutions did you explore? (Optional)
- NA

**Comment by VickyStash** [comment]
@GandalfGwaihir I've tried twice with your solution, and I still see the same problem 😅 But overall this part of the code was added in [this PR](https://github.com/Expensify/App/pull/33394/files), and it initially had this check for `reportIDFromParams`, so I haven't changed the logic.

**Comment by VickyStash** [comment]
@GandalfGwaihir Do you exactly follow the flow described in the ticket? You mention _deleted messages_, but in the test steps, it says to remove the workspace

**Comment by nkdengineer** [comment]
@aimane-chnaif Do you have any feedback about my suggestion [here](https://github.com/Expensify/App/issues/38183#issuecomment-1993881256)?

**Comment by blimpich** [comment]
@GandalfGwaihir yeah if reverting their PR doesn't fix the problem then their PR isn't the issue. Even if changing that line fixes the behavior what we want is the _recent_ change that caused this to occur.

@nkdengineer same thing, the explanation of the root cause of the problem is incomplete without explaining which one of [these PRs](https://github.com/Expensify/App/issues/38162) caused the regression.

I have a short work day today but I will see if I can find time to figure out the PR that caused this. If i can't figure that out in a reasonable amount of time I'm open to making this open to other contributors, but no proposal is complete without a root cause that identifies the offending PR.

**Comment by VickyStash** [comment]
@blimpich I'm still investigating the issue. I have found some clues, but not the root cause/offending PR yet. Let me know if I should continue my investigations tomorrow, I'll be happy to help

**Comment by VickyStash** [comment]
@blimpich @aimane-chnaif I've noticed there is an [open issue](https://github.com/Expensify/App/issues/37551) with [open PR](https://github.com/Expensify/App/pull/38251), which will block thread functionality for archived rooms. After merging that PR this issue will not be relevant anymore

**Comment by blimpich** [comment]
@VickyStash doesn't that PR just prevent creating new replies? You can still navigate to existing replies with the changes made in that PR I think, which would still leave us with this issue no?

**Comment by VickyStash** [comment]
@blimpich Here are my test results from [that PR](https://github.com/Expensify/App/pull/38251) (I switched and tested on it):
1. If the message in the archived chat has no thread, the tread button is not visible at all, and you can't get to the tread anymore (it covers this ticket test flow).
2. If the message has a thread and then the chat got archived, it displays without problem (it's so on the current main as well)
See the recording

https://github.com/Expensify/App/assets/23176449/6d538337-f813-4861-a033-c24eb2e0b28b

**Comment by blimpich** [comment]
Ah, cool! Thank you for the explanation! I edited the their PR description to include this issue in the list of issues that will be fixed when they merge. Also [posted a comment](https://github.com/Expensify/App/pull/38251#issuecomment-1997962137). Let's keep this open until (hopefully) Melvin auto-closes it.

**Comment by blimpich** [comment]
not overdue

**Comment by miljakljajic** [comment]
still waiting on aimaine

**Comment by miljakljajic** [comment]
Going to close this out and we can re-open when Aimane comes back. I have also DMed them in slack.

**Comment by mallenexpensify** [comment]
@aimane-chnaif can you please accept the job below? Please reply here and tag me once you have. 
https://www.upwork.com/jobs/~021988040412935524657
